shared/src/main/java/uk/gov/di/authentication/shared/matchers/AuditMessageMatcher.java

```diff
@@ -0,0 +1,70 @@
+package uk.gov.di.authentication.shared.matchers;
+
+import org.hamcrest.Description;
+import org.hamcrest.TypeSafeDiagnosingMatcher;
+import uk.gov.di.audit.AuditPayload;
+
+import java.nio.charset.StandardCharsets;
+import java.util.function.Function;
+
+public class AuditMessageMatcher<T> extends TypeSafeDiagnosingMatcher<String> {
+
+    private final String name;
+    private final Function<String, T> mapper;
+    private final T expected;
+
+    private AuditMessageMatcher(
+            String name, Function<AuditPayload.AuditEvent, T> mapper, T expected) {
+        this.name = name;
+        this.expected = expected;
+
+        Function<String, AuditPayload.AuditEvent>
+                annoyingInterimVariableBecauseFPIsANightmareInJava = this::deserialiseAuditEvent;
+        this.mapper = annoyingInterimVariableBecauseFPIsANightmareInJava.andThen(mapper);
```

Comment:
Probably the best replacement for this would be to avoid function composition entirely, for readability

```java
this.mapper = input -> mapper.apply(deserialiseAuditEvent(input));
```

Reply:
Ohhhhhh, yeah, that makes sense.  I kind of ended up looking at it backwards, for various reasons.